### PR TITLE
Pass tests on MSVC

### PR DESCRIPTION
**Pull request by eiennohito**
A path to #31 is almost over?

### CODE REVIEW

**Comment by eiennohito** [comment]
But the development on Windows is pain :(

**Comment by DoumanAsh** [comment]
>But the development on Windows is pain :(

Yeah, it is true in some way as it requires lot of setup.
But it is much more easier with new 2017.

I'll check out myself also clang with windows headers which would be also nice to support as it is much better compiler than `cl`

File: src/rnn/legacy/rnnlmlib_static.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ void CRnnLM_stat::ReadFRHeader(FILE *file) { /*{{{*/
   }
 
   // direct
-  unsigned long int maxent_hash_size;
+	int64_t maxent_hash_size;
```
Comment (DoumanAsh) [inline src/rnn/legacy/rnnlmlib_static.cpp]: correct tabulation? i now it is old code but still

File: src/core/core_version.h.in [hunk 2]
```diff
@@ -7,9 +7,9 @@
 namespace jumanpp {
 namespace core {
 
-constexpr const StringPiece JPP_VERSION_STRING { "@JUMANPP_FULL_VERSION@" };
+static const StringPiece JPP_VERSION_STRING { "@JUMANPP_FULL_VERSION@" };
```
Comment (DoumanAsh) [inline src/core/core_version.h.in]: I'm curious what is error here?

**Comment by eiennohito** [comment]
It's not the setup which is bad (MSVC/Msys2/Mingw are easy to install), but the compilation times (debug linking times on Mingw are awful), MSVC builds are slow as hell as well (I have a pretty good development machine). MSVC bugs? or quirks are really funny sometimes.

Btw, what are you trying to make? Maybe I can give you some advice :p

File: src/rnn/legacy/rnnlmlib_static.cpp (hunk 1 above)
Comment (eiennohito) [reply src/rnn/legacy/rnnlmlib_static.cpp]: I think its tabs vs spaces (boo on you, Visual Studio), and I couldn't get git-clang-format script working on Mingw (I haven't tried much though).

File: src/core/core_version.h.in (hunk 2 above)
Comment (eiennohito) [reply src/core/core_version.h.in]: This was a fun one.

A string literal constructor of a StringPiece (a one that is for const char &array[N]) was giving absurd results when used as a constexpr on MSVC. I thought that the easiest way was to get rid of it on that platform.

**Comment by DoumanAsh** [comment]
@eiennohito Hm as I'm using powershell and cmake with ninja I don't have slow times even with MSVC, but maybe it is because I'm doing stuff on gaming machine.

The thing I want to make is simple parser of japanese sentence in order to help myself with reading japanese (while I know grammar, there ain't much kanji I know).
There are some tools that exist for such purpose but I just wanted to try to make my own after seeing that jumanpp employs machine learning which may yield much better result than mecab.
They goal is basically take sentence(from clipboard) and present user with its components(plus using some dictionary to explain meaning of kanji)

File: src/core/core_version.h.in (hunk 2 above)
Comment (DoumanAsh) [reply src/core/core_version.h.in]: Now that's really sound like a MSVC bug...
It is just doesn't look nice since static makes it not thread safe

**Comment by DoumanAsh** [comment]
Small question: I'm trying to build your win branch currently but hit `..\src\rnn/simple_rnn_impl.h(8,10):  fatal error: 'eigen3/Eigen/Core' file not found`

Did you fixed it too ? I wonder why I hit it.

**Comment by eiennohito** [comment]
Just run the `dependencies.sh` script in the root folder for the time being.
Any unix-ish environment (e.g. Msys2 or WSL) should work.

I was using mostly Win before coming here in 2013.
About the compilation speed, I have an i7-6850 with hugeload of DDR4 (and GPUs for deep learning) desktop I use as a local development machine, and gcc compilation speed on Windows (Msys2) is ~2x slower than on Linux (maybe I have a bias though). No sanitizers on Win as well. 
Visual Studio debugger is really good though.

The VS problem probably is the lack of parallel compilation inside projects using the default CMake settings, so that may be solvable.

Btw my "main" research is actually automatic acquisition of high-quality example sentences for language learning. Basically - find nice sentences that try to explain the word in context. Jumanpp is a small-that-became-not-that-small detour basically. 

Probably, the worst problem with present morphological analyzers in relation to language/word learning is that you have to somehow bind several word segmentation standards together. E.g. you want to use JMdict as a definition dictionary, but the mapping between segmentation dictionaries and JMdict is many-to-many and it suck sometimes.

**Comment by eiennohito** [comment]
Ah, and don't learn kanji, learn words instead. :p
Kanji are better to be remembered as something like sets of words which contain this kanji imho.

File: src/core/core_version.h.in (hunk 2 above)
Comment (eiennohito) [reply src/core/core_version.h.in]: Constants are always thread safe. I agree that constexpr would be better, but static const is also ok.

And you are mistaken about the meaning of static here: it basically means that each compilation unit gets its own copy of the thing and can treat it as a local (non-shared) object. static has too much meanings in C++.
Usually that makes it possible to the compiler to inline things and do some other optimizations.

**Comment by DoumanAsh** [comment]
The problem is that currently I have no unix environment... (I'll try to look into my git installation)
Requiring unix environment maybe problematic... but I guess solvable.

>The VS problem probably is the lack of parallel compilation inside projects using the default CMake settings, so that may be solvable.

I guess yes, I'm using ninja as it is able to parallel builds

>Btw my "main" research is actually automatic acquisition of high-quality example sentences for language learning.
Do you mean jumanpp?
I guess most research materials are written in japanese?
One of the reasons though I'm trying to jumanpp is japanese too context dependent so machine learning should really improve tokenization, or at least I hope.

>Ah, and don't learn kanji, learn words instead. :p
>Kanji are better to be remembered as something like sets of words which contain this kanji imho.

Yes, I know. But in order to know words you need to know kanji with some common readings :)
So it is helpful when you read with parser to speed up process of reading also as I'm not just reading but playing games :)

**Comment by eiennohito** [comment]
I strongly recommend Msys2 as a lightweight unix environment on Windows. Mingw (which is msys's brother-like something) is actually more like using gcc with winapi without unix-y stuff.

Mecab is not *bad* at tokenization actually. And it is blazing fast because of the architecture (but there are drawbacks of that). It just does its share of stupid mistakes.

About readings: for me it was better like: I know the reading of a kanji because I know its reading in a couple of words. I really couldn't remember the reading of plain kanji because there was nothing to bind that reading to. So words, you bind meanings, readings and kanji to them and slowly expand on the whole language.

I mean https://kotonoha.ws/ and http://www.aclweb.org/anthology/W17-5014 (the backend when its working).
I write my stuff in English because I'm lazy. :p

**Comment by DoumanAsh** [comment]
>I strongly recommend Msys2 as a lightweight unix environment on Windows. Mingw (which is msys's brother-like something) is actually more like using gcc with winapi without unix-y stuff.

Few years ago I was using it and I know it is pretty cool.
But currently I switched fully to MSVC as it produces better binaries and plus I can utilize powershell fully for my development.

>Mecab is not bad at tokenization actually. And it is blazing fast because of the architecture (but there are drawbacks of that). It just does its share of stupid mistakes.

Yes, I actually tried it and had some simple wrapper over it.
The only problem is that it is not exactly unicode aware so initially I had some troubles.
Well aside from rather stupid way to load dictionary and inability to use static linkage.

>About readings: for me it was better like: I know the reading of a kanji because I know its reading in a couple of words.
Yes, I noticed it too. I'm currently learning kanji with Andrew Scott Conning - The Kodansha Kanji Learner's Course
While some can learned without associated words, it is of course easier to remember kanji which is part of some word you know.
But I'm trying to do it in parallel.

I'll take a look at links. Kotonoha seems to be useful tool :)

>I write my stuff in English because I'm lazy. :p
English is that good language :D

**Comment by eiennohito** [comment]
MSVC binaries were actually smaller, yep. But I wonder what about the execution speed. On Linux gcc produced the fastest binaries for me (even faster than Intel compiler).

You need to build your fully utf-8 dictionary/model for MeCab to make it utf-8-ish.

Btw Jumanpp won't support *dynamic* linkage for the time being I think :p

File: src/core/core_version.h.in (hunk 2 above)
Comment (DoumanAsh) [reply src/core/core_version.h.in]: Ah my bad, i forgot that we're talking about header here.

**Comment by DoumanAsh** [comment]
MSVC is likely behind GCC in speed alone it is true, but then it is most likely due to better SSE and instincts supports.
But I don't have definite numbers.

>Btw Jumanpp won't support dynamic linkage for the time being I think :p
I'm all for static linkage :D

**Comment by eiennohito** [comment]
Ok, finally took my hand to do correct formatting.

Windows build should be usable after the merge.

**Comment by DoumanAsh** [comment]
Thanks!

**Pull request merged**